### PR TITLE
GCP deprovision: Plumb NetworkProjectID for shared VPC

### PR DESCRIPTION
**Pull request by 2uasimojo**
The shared VPC flow added the need for the installer to use a NetworkProjectID to manage certain resources. This is provided at installation time via the install-config, and is included in the metadata.json produced by the installer when provisioning completes.

The uninstall process requires this NetworkProjectID during deprovisioning to perform the corresponding cleanup -- without it, those resources are leaked. So we need to plumb that datum through from the metadata.json to the uninstall process.

To make this more general-purpose for adoption and disaster recovery, and for retrofitting existing clusters that were created before this commit, we also attempt to pull the NetworkProjectID out of the install-config if it is unset.

NOTE: We will refuse to proceed with deprovision if this field is not set. Once this change lands, we should audit existing clusters to make sure they all get retrofitted; and manually retrofit any others (with the empty string).

[HIVE-2334](https://issues.redhat.com//browse/HIVE-2334)

### CODE REVIEW

**Comment by 2uasimojo** [comment]
/assign @lleshchi 

This very closely follows the example of #2100.

File: pkg/controller/clusterdeployment/clusterprovisions.go [hunk 1]
```diff
@@ -367,6 +368,38 @@ func setAzureResourceGroupFromMetadata(cd *hivev1.ClusterDeployment, cm *hivev1.
...
+	if err := json.Unmarshal(pmjson, &im); err != nil {
+		logger.WithError(err).Error("Could not unmarshal ClusterMetadata!")
+		return
+	}
+
+	if im.GCP == nil {
+		logger.Warn("ClusterMetadata unexpectedly has no GCP section")
+		return
+	}
+	npid := im.GCP.NetworkProjectID
```
Comment (lleshchi) [inline pkg/controller/clusterdeployment/clusterprovisions.go:389]: Does this section require a check for whether `im.GCP.NetworkProjectID` is not nil, which would be different from an empty string? (Assuming it's not necessary but would like to know why)

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 2]
```diff
@@ -2466,6 +2467,62 @@ func (r *ReconcileClusterDeployment) discoverAzureResourceGroup(cd *hivev1.Clust
...
+// discoverGCPNetworkProjectID is intended to retrofit GCP clusters which were a) adopted, or b) created before
+// support for shared VPC was added. It attempts to discover the network project ID for the cluster by parsing
+// the install-config.
+// No op if platform is not GCP.
+// No op if the field is already set (we don't validate that it is correct).
+// This is best effort. If things go wrong, we log but do not fail out. THIS WILL BLOCK DEPROVISIONING. If
+// this happens, the field must be set manually.
+// The return indicates whether the field was modified. If `true`, the caller is responsible for pushing
+// an Update() back to the server.
+func (r *ReconcileClusterDeployment) discoverGCPNetworkProjectID(cd *hivev1.ClusterDeployment, log log.FieldLogger) bool {
```
Comment (lleshchi) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go:2479]: Where is this function used?

File: pkg/controller/clusterdeployment/clusterprovisions.go (hunk 1 above)
Comment (2uasimojo) [reply pkg/controller/clusterdeployment/clusterprovisions.go:389]: Is the field in installertypes.ClusterMetadata a pointer? I thought it was a string type.

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 2 above)
Comment (2uasimojo) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:2479]: Did I forget the bit where it gets called? :sheepish:

File: pkg/controller/clusterdeployment/clusterprovisions.go (hunk 1 above)
Comment (lleshchi) [reply pkg/controller/clusterdeployment/clusterprovisions.go:389]: Oh, you're right. Makes sense.

Comment (2uasimojo) [reply pkg/controller/clusterdeployment/clusterprovisions.go:389]: [ref](https://github.com/openshift/hive/blob/a9aaf72f8d1849aa5b5f2e3d6e137e93d9fbbdb5/vendor/github.com/openshift/installer/pkg/types/gcp/platform.go#L26)

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 2 above)
Comment (2uasimojo) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:2479]: Fixed.

**Pull request merged**